### ISSUE TITLE
lint request: enforce recommended dartdoc structure

### ISSUE THREAD

**Issue by pq** [dart-lang/linter#1433]
In https://github.com/flutter/flutter/pull/27791#discussion_r255772365, @goderbauer wrote:

> That a dartdoc should start with a one-sentence summary as its own paragraph is something I have to frequently point out in code reviews - so I'd prefer it if things like these could be pointed out automatically.

👍 

Let's talk details.

1. Enforcing a summary statement lines up w/ this advice from Effective Dart and I think we should just do it: https://www.dartlang.org/guides/language/effective-dart/documentation#do-separate-the-first-sentence-of-a-doc-comment-into-its-own-paragraph

Other thoughts?

**Comment by goderbauer** [comment]
Yes, yes, yes! Please!

**Comment by srawlins** [comment]
One thing I'd like to guard against: Pythnon's linter (one of them) enforces a one _line_ summary which I think is a disaster. It limits the summary to something like 75 characters, yuck. One _sentence_ is the Effective Dart recommendation, and we should make sure to stick to that.

I'm not volunteering to write the regex / Markdown parser for this 😁

**Comment by pq** [comment]
> I'm not volunteering to write the regex / Markdown parser for this 😁

Hopefully we can just use the parser from [`package:markdown`](https://github.com/dart-lang/markdown) for this?  Alternatively we could crib some bits from `dartdoc`?

**Comment by pq** [comment]
Or is that even over-thinking it?  Can we just get away with something simple like

```dart
comment.split('\n\n');
```

?

**Comment by goderbauer** [comment]
How would you make sure that after splitting the first part is just one sentence though?

**Comment by pq** [comment]
Yeah.  I realized that.  A simple scanner might work but then maybe there are complex cases.

@jcollins-g: would it be easy for us to crib this logic from `dartdoc`?

**Comment by jcollins-g** [comment]
There is logic here in dartdoc that divides the HTML output into two segments:

https://github.com/dart-lang/dartdoc/blob/9e5a57dd1d0826d5a59fd074d8eb2637466a5c17/lib/src/markdown_processor.dart#L874

As @pq surmised, we take the first node of generated HTML for the summary (modulo a bunch of dartdoc special casing and stripping of script tags, etc).  We discussed offline and my recommendation for lints that are likely to produce decent results in dartdoc regardless of localization or personal preference would be:

- Make sure the comment starts with text (no code blocks right at the beginning, or script tags, or raw HTML, etc).
- Make sure the first paragraph break happens in two lines or less.

I think narrowing it down to a "one sentence" summary is probably less general.  Maybe a separate lint not enabled by default if we really want it?

**Comment by pq** [comment]
After chatting w/ @jcollins-g and @bwilkerson, I'm wondering if the following might not get us a lot of the way there:

1.  a lint that flags docs whose summary paragraph is longer than some reasonable length.  Notice how `MethodInvocation` wraps below.  We could flag that and suggest they break the doc into summary and details.

    ![image](https://user-images.githubusercontent.com/67586/52657806-4f3b8a00-2eae-11e9-93c1-e2665f98fe92.png)

2. a lint that flags doc summaries that contain stuff that wouldn't render well (e.g., doesn't start w/ text, maybe contains links?)


@goderbauer: thoughts?

**Comment by goderbauer** [comment]
I've never really seen 2 as a problem, so not sure how useful that one will be...

For 1 I'd be curious what we'd consider a reasonable length to not run into Python's annoying limitation of "this has to be 80 chars or less"...

ICU has some heuristics to detect sentence breaks based on Unicode recommendation: http://userguide.icu-project.org/boundaryanalysis#TOC-Sentence-Boundary Maybe that can be useful?

**Comment by pq** [comment]
> I've never really seen 2 as a problem, so not sure how useful that one will be...

I'm not sure either to be honest but then the Flutter docs have benefited from extraordinary care.  This may be more valuable to more slap-dash efforts.

> For 1 I'd be curious what we'd consider a reasonable length to not run into Python's annoying limitation of "this has to be 80 chars or less"...

Me too.  @jcollins-g: I'm curious if you have a gut sense for this?

Thanks for the ICU link.  It would be interesting to see how many false positives we'd get with their heuristics.  Would that there was a dart package implementation so we could do some quick testing!

**Comment by jcollins-g** [comment]
> > I've never really seen 2 as a problem, so not sure how useful that one will be...
> 
> I'm not sure either to be honest but then the Flutter docs have benefited from extraordinary care. This may be more valuable to more slap-dash efforts.
> 
> > For 1 I'd be curious what we'd consider a reasonable length to not run into Python's annoying limitation of "this has to be 80 chars or less"...
> 
> Me too. @jcollins-g: I'm curious if you have a gut sense for this?

As above, my gut sense is "two lines" -- of any length.  ~150 characters might also work.

> 
> Thanks for the ICU link. It would be interesting to see how many false positives we'd get with their heuristics. Would that there was a dart package implementation so we could do some quick testing!

**Comment by ditman** [comment]
I want this sooo badly. Can it please also enforce that the first line **ends in a "."**?

**Comment by srawlins** [comment]
Don't expect this any time soon.

A rule was attempted in https://github.com/dart-lang/linter/pull/2148 and was deemed too slow. The comments over there highlight a lot of the dragons to be found in an implementation. All of that being said, contributions welcome! I'd love this rule too :D. This is one of the most-referenced Effective Dart rules in internal readability reviews.

**Comment by ditman** [comment]
@srawlins that rule seems way more complex than what I'd be (very) happy with ((I'm a simple man :P)).

I'm not suggesting to validate the whole doc block to add periods at the end of each paragraph or anything, just that the summary line of the DartDoc comment (the first/only one) does :P

Something like this:

* https://checkstyle.sourceforge.io/config_javadoc.html#SummaryJavadoc (but with even fewer features!)

I wouldn't even bother with the *length* of the line, unless very long lines end up looking terrible in the processed DartDoc output!

(Checkstyle allows to configure the "period" character at the end, which is cool for programmers writing comments in japanese, though)

_((PS: This looks interesting, I might give it a shot!))_

**Comment by srawlins** [comment]
I think that is a separate request from this one. The Effective Dart guideline requested here is: https://www.dartlang.org/guides/language/effective-dart/documentation#do-separate-the-first-sentence-of-a-doc-comment-into-its-own-paragraph

You might be looking for https://github.com/dart-lang/linter/issues/2147 or https://github.com/dart-lang/linter/issues/56.
